### PR TITLE
feat: Ability to unrelate private documents from actors

### LINKED ISSUE #2906: Revoke or unshare a document previously created a relation for
Revoke or unshare a document previously created a relation for

### PR DESCRIPTION
**Pull request by shahzadlone**
## Relevant issue(s)
Resolves #2906 

## Description
Follow-up to https://github.com/sourcenetwork/defradb/pull/2907, the ability to delete a relationship (in order to revoke access from an identity).

## CLI Demo
The following revokes the target actors 'relational' access to the private doc (can't read anymore): 

```bash
defradb client acp relationship delete \
--collection Users \
--docID bae-ff3ceb1c-b5c0-5e86-a024-dd1b16a4261c \
--relation reader \
--actor did:key:z7r8os2G88XXBNBTLj3kFR5rzUJ4VAesbX7PgsA68ak9B5RYcXF5EZEmjRzzinZndPSSwujXb4XKHG6vmKEFG6ZfsfcQn \
--identity e3b722906ee4e56368f581cd8b18ab0f48af1ea53e635e3f7b8acd076676f6ac
```

Result:
```json
{
  "RecordFound": true // <-------------- Indicates a relationship was found and deleted
}
```

## How has this been tested?
 CI + Integration Tests + Unit Tests

### CODE REVIEW

File: http/handler_acp.go
Comment (AndrewSisley) [inline http/handler_acp.go:147]: todo: This looks incorrect, you should be strongly typing the schema so that the openAPI magic can document it properly.  Have a look at `openApiSchemas` in http/openAPI.go

Please test this manually after correcting.

File: tests/integration/acp.go
Comment (AndrewSisley) [inline tests/integration/acp.go:417]: question: Why are these requires wrapped in an if, and not just done after the calls to `AssertError` and `assertExpectedErrorRaised`?

File: tests/integration/acp.go
Comment (AndrewSisley) [inline tests/integration/acp.go:364]: todo: This seems like an unmanagably large if-else block and it looks like there is a fair amount of duplication.  Please rework this function.

**Review by AndrewSisley — CHANGES_REQUESTED** [review]
It looks good, but there are a couple of important todos for you before I can approve.  My brain is a little foggy atm too, so it might be wise to wait for someone else to review too.

File: http/handler_acp.go
Comment (shahzadlone) [reply http/handler_acp.go]: Resolved in commit: `PR(ADHERE-REVIEW): Fix the open api schemas`

File: tests/integration/acp.go
Comment (shahzadlone) [reply tests/integration/acp.go:364]: This is an issue since the previous node index bug was fixed, I will be taking care of this in https://github.com/sourcenetwork/defradb/issues/3076

Won't do in this PR.

File: tests/integration/acp.go
Comment (shahzadlone) [reply tests/integration/acp.go:417]: Is copy pasted from `addPolicy` and the `addDocActorRelationshipACP`, I don't remember of the top of my head.

Comment (AndrewSisley) [reply tests/integration/acp.go:417]: Can you change both then please?

File: tests/integration/acp.go
Comment (AndrewSisley) [reply tests/integration/acp.go:364]: If you really don't want to tidy it up a bit in this function can you please add a todo comment linked to that ticket and make sure this function is mentioned in the issue?

Comment (shahzadlone) [reply tests/integration/acp.go:364]: > If you really don't want to tidy it up a bit in this function can you please add a todo comment linked to that ticket and make sure this function is mentioned in the issue?

I can if you really like, but I literally have the fix I am working on in a separate branch so will be opening the clean up PR soon

Comment (AndrewSisley) [reply tests/integration/acp.go:364]: Ah okay, if it is all coded up already somewhere nearly ready to go dont bother here :)

Comment (shahzadlone) [reply tests/integration/acp.go:364]: Cheers

File: tests/integration/acp.go
Comment (shahzadlone) [reply tests/integration/acp.go:417]: > Can you change both then please?

Sure, I don't mind :)

Comment (shahzadlone) [reply tests/integration/acp.go:417]: - [ ] Remove `err == nil` if check from test

Comment (shahzadlone) [reply tests/integration/acp.go:417]: Update: I did remove the asserts in the end and the tests started failing it's because it does the checks weather an error was raised or not. I am undoing the change. If I can remember I did do this change because It would upon failure sometimes show a false positive error not raised. I am too lazy to understand why haha but the current framework works, so keeping as is

Comment (AndrewSisley) [reply tests/integration/acp.go:417]: I would guess that you forgot to return/continue early if `expectedErrorRaised == true` and continued to the value assertion after receiving the expected error

**Review by AndrewSisley — APPROVED** [review]
Looks good to me, but I would prefer you got a second reviewer to look over this one before merge :) (don't wait more than a couple of working days for that though)

File: cli/acp_relationship_delete.go
Comment (nasdf) [inline cli/acp_relationship_delete.go:115]: todo: short flag is empty

File: http/handler_acp.go
Comment (nasdf) [inline http/handler_acp.go:81]: suggestion: panic or internal server error (500) would be better imo. Bad request means the user did something wrong.

**Review by nasdf — APPROVED** [review]
Two minor things to consider before merging. The documentation and tests are awesome as always!

File: cli/acp_relationship_delete.go
Comment (shahzadlone) [reply cli/acp_relationship_delete.go:115]: IIRC I don't think we ever had a short flag for --docID

What short flag are you suggesting?

File: tests/integration/acp.go
Comment (shahzadlone) [reply tests/integration/acp.go:417]: But then that's still another if check anyways. I am happy with the current setup. Is fairly easy to change later for all the test actions if we want.

Comment (shahzadlone) [reply tests/integration/acp.go:417]: But then that's still another if check anyways. I am happy with the current setup. Is fairly easy to change later for all the test actions if we want.

Comment (AndrewSisley) [reply tests/integration/acp.go:417]: Another `if` yes, but you are no longer spoofing the main error assertion logic.

File: cli/acp_relationship_delete.go
Comment (nasdf) [reply cli/acp_relationship_delete.go:115]: I would replace it with `StringVar` if there's no shorthand.

Comment (shahzadlone) [reply cli/acp_relationship_delete.go:115]: Ah, I see that is what you meant. Will do.

- [ ] Use `StringVar` for `--docID`

Comment (shahzadlone) [reply cli/acp_relationship_delete.go:115]: Resolved in commit: ```PR(ADHERE-REVIEW): Use `StringVar` for only long flag arg```

File: http/handler_acp.go
Comment (shahzadlone) [reply http/handler_acp.go:81]: Discussed on discord will resolve in:
https://github.com/sourcenetwork/defradb/issues/3103

File: tests/integration/acp.go
Comment (shahzadlone) [reply tests/integration/acp.go]: Resolved in commit: ``` PR(ADHERE-REVIEW): Check `require.Equal` after the `assertExpectedErrorRaised` ```

**Pull request merged**
